decidim-core/lib/tasks/upgrade/decidim_fix_nickname_uniqueness.rake

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+namespace :decidim do
+  namespace :upgrade do
+    desc "Modify nicknames when exists similar ones case insensitively"
+    task fix_nickname_uniqueness: :environment do
+      logger = Logger.new($stdout)
+      logger.info("Updating users nickname ...")
+
+      # list of users already changed in the process
+      has_changed = []
+
+      Decidim::User.find_each do |user|
+        next if has_changed.include? user
+
+        Decidim::User.where("nickname ILIKE ?", user.nickname.downcase).order(:created_at).each do |similar_user|
+          next if has_changed.include? similar_user
+          next if user == similar_user
+
+          # change her nickname to the lowercased one with 5 random numbers
+          begin
+            update_user_nickname(similar_user, "#{similar_user.nickname}-#{rand(99_999)}")
+          rescue ActiveRecord::RecordInvalid => e
+            logger.warn("User ID (#{similar_user.id}) : #{e}")
+            update_user_nickname(similar_user, "#{similar_user.nickname}-#{rand(99_999)}")
```

Comment:
I don't like using the `update_user_nickname` again in the rescue clause. If it has failed the first time, it'll probably fail the second time too. 

How about adding the users that had given error to an array and adding it conditionally to the "Process terminated" message? Something like this (feel free to change it, of course): 

```ruby
     if users_with_errors.zero?
       logger.info("Process terminated successfully, #{has_changed.count} users nickname have been updated.")
    else
       logger.info("Process terminated with errors")
       logger.info("#{has_changed.count} users nickname have been updated successfully. 
       logger.info("#{users_with_errors.count} users had errors. Please check them manually in the console to solve it")
       logger.info("User IDs with errors: #{users_with_errors.pluck(:id)}")
     end
```

What do you think @eliegaboriau?

Reply:
We added this in the rescue clause thinking that it would failed if the random number generate a nickname already taken. By retrying it, it has almost no chance of failing again for this reason.
I don't really know if it could fail for any other reason, maybe I can implement your proposal and add an advice to re-run the task if some users have failed, so the users failing with the random can be updated successfully ?

Reply:
We already have a method for this which is also used for the omniauth registrations here:
https://github.com/decidim/decidim/blob/833bb0a52e2ada24714b04dfe396691a15c555f3/decidim-core/app/forms/decidim/omniauth_registration_form.rb#L28

So, instead of generating the username manually here, I believe you can do this:
```suggestion
            update_user_nickname(similar_user, Decidim::UserBaseEntity.nicknamize(similar_user.nickname))
```

The format is slightly different (underscore instead of dash) but I believe it does exactly what you are looking to do here.

After you do this, you can get rid of the `begin..rescue` block completely and make @andreslucena happier too.

Reply:
It should be ok then, i've let the rescue block but i don't know if it's really useful

Reply:
> i've let the rescue block but i don't know if it's really useful

I'd say you can keep the rescue block there now that it's only displaying a message to the console (which is useful if that situation happens) and not trying to update it again.

It should "never" go into the rescue block, as `2.step { |n| ... }` will continue the iteration until you break out from the block:
https://github.com/decidim/decidim/blob/2892dcf4ee492a0d494d6f83e897f78688388ac2/decidim-core/lib/decidim/nicknamizable.rb#L54

So, this iteration will continue to infinity or until it finds an available nickname. Which means that the only place where this can fail is that if it generates a nickname which is longer than the `nickname` column can hold in the database (which is 20 characters).

So theoretically, this can happen because 20 characters is not that much. If there are e.g. two users who want the exact same long username, this can happen. Or if the nicknames are auto-generated from the actual name (through omniauth login), it could happen if two people have the exact same long name.

---

By the way I just realized that we also need to do the lowercase search here too:
https://github.com/decidim/decidim/blob/2892dcf4ee492a0d494d6f83e897f78688388ac2/decidim-core/lib/decidim/nicknamizable.rb#L55

Could you also take care of that too please @eliegaboriau ?

Reply:
> By the way I just realized that we also need to do the lowercase search here too:

I've the impression it works without it, as it passes the tests here
https://github.com/decidim/decidim/blob/0026fb69d9323e3b43c5d627b1c9b54af010df50/decidim-core/spec/tasks/decidim_tasks_fix_nickname_uniqueness_spec.rb#L23-L27

Reply:
Nope, you're absolutely right, my bad

Reply:
Thanks @eliegaboriau, LGTM.

@andreslucena Does this make you happier now?

Reply:
Yes, I am! I didn't know about `nicknamize` really, much better and cleaner!